source-mongodb/pull.go

```diff
@@ -255,54 +295,123 @@ func (c *capture) ChangeStream(ctx context.Context, client *mongo.Client, bindin
+		return fmt.Errorf("could not run find query on collection %s: %w", res.Collection, err)
 	}
 
 	defer cursor.Close(ctx)
 
 	var i = 0
 	for cursor.Next(ctx) {
 		i += 1
+
+		// We decode once into a map of RawValues so we can keep the _id as a
+		// RawValue. This allows us to marshal it as an encoded byte array along
+		// with its type code in the checkpoint, and unmarshal it when we want to
+		// use it to resume the backfill, ensuring the type of the _id is consistent
+		// through this process
+		var rawDoc map[string]bson.RawValue
+		if err = cursor.Decode(&rawDoc); err != nil {
+			return fmt.Errorf("decoding document raw in collection %s: %w", res.Collection, err)
+		}
+		state.BackfillLastId = rawDoc[idProperty]
+
 		var doc bson.M
 		if err = cursor.Decode(&doc); err != nil {
-			return nil, fmt.Errorf("decoding document in collection %s: %w", res.Collection, err)
+			return fmt.Errorf("decoding document in collection %s: %w", res.Collection, err)
 		}
 		doc = sanitizeDocument(doc)
 
 		js, err := json.Marshal(doc)
 		if err != nil {
-			return nil, fmt.Errorf("encoding document %v in collection %s as json: %w", doc, res.Collection, err)
+			return fmt.Errorf("encoding document %v in collection %s as json: %w", doc, res.Collection, err)
 		}
 
 		if err = c.Output.Documents(binding, js); err != nil {
-			return nil, fmt.Errorf("output documents failed: %w", err)
+			return fmt.Errorf("output documents failed: %w", err)
 		}
 
 		if i%CHECKPOINT_EVERY == 0 {
-			if err = c.Output.Checkpoint([]byte("{}"), true); err != nil {
-				return nil, fmt.Errorf("output checkpoint failed: %w", err)
+			var checkpoint = captureState{
+				Resources: map[string]resourceState{
+					resourceId(res): *state,
+				},
+			}
+
+			checkpointJson, err := json.Marshal(checkpoint)
+			if err != nil {
+				return fmt.Errorf("encoding checkpoint to json failed: %w", err)
+			}
+
+			if err = c.Output.Checkpoint(checkpointJson, true); err != nil {
+				return fmt.Errorf("output checkpoint failed: %w", err)
+			}
+
+			if diff, err := OplogTimeDifference(ctx, client); err != nil {
+				return fmt.Errorf("could not read oplog time difference: %w", err)
+			} else {
+				log.WithField("timediff", diff).Debug("read oplog time difference")
+
+				// TODO: This is just spamming logs, but users won't see these logs. Should we
+				// just error after seeing this error for a while?
+				if diff < minOplogTimediff {
+					log.Warn(fmt.Sprintf("the current time difference between oldest and newest records in your oplog is %d seconds. This is smaller than the minimum of 24 hours. Please resize your oplog to be able to safely capture data from your database: https://go.estuary.dev/NurkrE", diff))
+				}
 			}
 		}
 	}
 
-	if err = c.Output.Checkpoint([]byte("{}"), true); err != nil {
-		return nil, fmt.Errorf("sending checkpoint failed: %w", err)
+	state.Status = StatusStreaming
```

Comment:
Do we need to check `cursor.Err` after `cursor.Next` returns `false`? Maybe not if an error would always be surfaced in `cursor.Decode`.

Reply:
We do it a few lines below

Reply:
Ah I missed that. But we emit the checkpoint first? That seems like it could (maybe) result in a case where we emit a checkpoint showing that the backfill is done if `cursor.Next()` returns false with an error, and the backfill isn't really done yet.

Reply:
@williamhbaker good point!